test/vtable_test.cpp

```diff
@@ -0,0 +1,129 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <iostream>
+#include <memory>
+
+#include <unifex/vtable.hpp>
+
+#include <gtest/gtest.h>
+
+struct simple_vtable_t {
+  UNIFEX_VTABLE_DECLARE;
+
+  UNIFEX_VTABLE_ENTRY_VOID(foo, int);
+  UNIFEX_VTABLE_ENTRY(bar, int, double, double);
+};
+
+struct class_one {
+  int do_foo() noexcept { return 1; }
+  int do_bar(double, double) { return 0; }
+
+  simple_vtable_t table_ =
+      UNIFEX_VTABLE_CONSTRUCT(&class_one::do_foo, &class_one::do_bar);
+};
+
+struct class_two {
+  explicit class_two(int v) : value_(v){};
+  class_two() = default;
+
+  int value_ = 2;
+
+  int foo_fn() noexcept { return value_; }
+  int bar_fn(double, double) { return 1; }
+
+  simple_vtable_t table_ =
+      UNIFEX_VTABLE_CONSTRUCT(&class_two::foo_fn, &class_two::bar_fn);
+};
+
+TEST(vtable, bar_returns_expected) {
+  auto instance_0 = std::make_unique<class_one>();
+  ASSERT_TRUE(!!instance_0->table_);
+  EXPECT_EQ(0, instance_0->table_.bar(0, 0));
+
+  auto instance_1 = std::make_unique<class_two>();
+  ASSERT_TRUE(!!instance_1->table_);
+  EXPECT_EQ(1, instance_1->table_.bar(0, 0));
+}
+
+TEST(vtable, foo_indirect_goes_to_instance) {
+  auto instance_0 = std::make_unique<class_one>();
+  auto instance_1 = std::make_unique<class_two>();
+  auto instance_2 = std::make_unique<class_two>(3);
+
+  simple_vtable_t* tables[] = {
+      &instance_0->table_, &instance_1->table_, &instance_2->table_};
+
+  for (auto& i : tables) {
+    ASSERT_TRUE(!!i);
+  }
+
+  EXPECT_EQ(tables[0]->foo(), 1);
+  EXPECT_EQ(tables[1]->foo(), 2);
+  EXPECT_EQ(tables[2]->foo(), 3);
+}
+
+struct reference_vtable_t {
+  UNIFEX_VTABLE_DECLARE;
+  UNIFEX_VTABLE_ENTRY(foo, std::string, std::string&&);
+  UNIFEX_VTABLE_ENTRY(bar, std::string, std::string&);
+};
+
+struct class_three {
+  std::string foo(std::string&& s) { return std::move(s); };
+  std::string bar(std::string& s) {
+    auto r = s;
+    s = "";
+    return r;
+  };
+
+  reference_vtable_t table_ =
+      UNIFEX_VTABLE_CONSTRUCT(&class_three::foo, &class_three::bar);
+};
+
+TEST(vtable, perfect_forwarding) {
+  auto instance = std::make_unique<class_three>();
+  ASSERT_TRUE(!!instance->table_);
+
+  static constexpr auto a = "string a";
+  std::string ao = a;
+  auto ar = instance->table_.foo(std::move(ao));
+  EXPECT_EQ(a, ar);
+
+  static constexpr auto b = "string b";
+  std::string bo = b;
+  auto br = instance->table_.bar(bo);
+  EXPECT_EQ(b, br);
+  EXPECT_EQ("", bo);
+}
+
+struct class_four {
+  struct vtable {
+    UNIFEX_VTABLE_DECLARE;
+    UNIFEX_VTABLE_ENTRY_VOID_RVALUE(foo, void);
+    UNIFEX_VTABLE_ENTRY_RVALUE(bar, void, int);
+  };
+
+  void foo() && {};
+  void bar(int) && {};
+
+  vtable table_ = UNIFEX_VTABLE_CONSTRUCT(&class_four::foo, &class_four::bar);
+};
+
+TEST(vtable, ref_qualifier) {
+  auto instance = std::make_unique<class_four>();
+  std::move(instance->table_).foo();
```

Comment:
Does it have to be `std::move()`? Can it be called on a (const?) reference? The vtable state does not change as a result of the call.

Besides, current design seems to imply that a vtable is always assigned to `this` object, and the table is a part of the class definition. Is such limitation desired? Can I define a vtable for `class_four` outside of `class_four`?

Reply:
> Does it have to be std::move()? Can it be called on a (const?) reference?

This code was attempting to show that ref-qualified functions can also be added to a vtable.  I'm pretty sure you can call on any r-value reference'd instance.  I could add an example on a local temporary here as well.

> The vtable state does not change as a result of the call.

The vtable is really just an indirection to the object itself.  In this case the object is needing to be called with r-value semantics (due to the ref-qualifier on the function), so the vtable does as well.  Otherwise the vtable code would be silently turning `this` into an r-value ref, which is dangerous.

> Besides, current design seems to imply that a vtable is always assigned to this object, and the table is a part of the class definition. Is such limitation desired? Can I define a vtable for class_four outside of class_four?

The vtable definition is allowed to be outside the class, and would be in most cases.  See lines 23 and 78.  `class_four` is the only example where I did it _inside_ the class, partially to show that it could be done there as well and partially for the implementation of the ref-qualifier test to be smaller.

The vtable instance _must_ be in the class (using the UNIFEX_VTABLE_CONSTRUCT call) just like a normal vtable would be, but the definition can and most likely will be outside the class.  This is similar to a "base class" with normal virtualization.